kuka_iiqka_eac_driver/src/hardware_interface.cpp

```diff
@@ -228,135 +182,74 @@ CallbackReturn KukaEACHardwareInterface::on_configure(const rclcpp_lifecycle::St
-  observe_thread_ = std::thread(&KukaEACHardwareInterface::ObserveControl, this);
-
-  OpenControlChannelRequest request;
-  OpenControlChannelResponse response;
-  grpc::ClientContext context;
+  prev_control_mode_ = static_cast<kuka_drivers_core::ControlMode>(hw_control_mode_command_);
 
-  request.set_ip_address(info_.hardware_parameters.at("client_ip"));
-  request.set_timeout(5000);
-  request.set_cycle_time(4);
-  request.set_external_control_mode(
-    kuka::motion::external::ExternalControlMode(hw_control_mode_command_));
   RCLCPP_INFO(
-    rclcpp::get_logger("KukaEACHardwareInterface"), "Starting control in %s",
-    kuka::motion::external::ExternalControlMode_Name(static_cast<int>(hw_control_mode_command_))
-      .c_str());
-
-  auto ret = stub_->OpenControlChannel(&context, request, &response);
-  if (ret.error_code() != grpc::StatusCode::OK)
-  {
-    RCLCPP_ERROR(rclcpp::get_logger("KukaEACHardwareInterface"), "%s", ret.error_message().c_str());
-    return CallbackReturn::FAILURE;
-  }
-#endif
+    rclcpp::get_logger("KukaEACHardwareInterface"),
+    "External control session started successfully");
 
+  stop_requested_ = false;
   return CallbackReturn::SUCCESS;
 }
 
 CallbackReturn KukaEACHardwareInterface::on_deactivate(const rclcpp_lifecycle::State &)
 {
-  RCLCPP_INFO(rclcpp::get_logger("KukaEACHardwareInterface"), "Deactivating");
+  RCLCPP_INFO(rclcpp::get_logger("KukaEACHardwareInterface"), "Deactivating hardware interface");
 
-  control_signal_ext_.control_signal.stop_ipo = true;
-  while (is_active_)
-  {
-    std::this_thread::sleep_for(std::chrono::milliseconds(10));
-  }
+  stop_requested_ = true;
 
-#ifdef NON_MOCK_SETUP
-  if (context_ != nullptr)
-  {
-    context_->TryCancel();
-  }
-#endif
-  if (observe_thread_.joinable())
-  {
-    observe_thread_.join();
-  }
   return CallbackReturn::SUCCESS;
 }
 
 return_type KukaEACHardwareInterface::read(const rclcpp::Time &, const rclcpp::Duration &)
 {
-#ifndef NON_MOCK_SETUP
-  std::this_thread::sleep_for(std::chrono::microseconds(3900));
-  for (size_t i = 0; i < info_.joints.size(); i++)
-  {
-    hw_position_states_[i] = hw_position_commands_[i];
-  }
-  return return_type::OK;
-#endif
+  // Bigger timeout blocks controller configuration
+  kuka::external::control::OperationStatus receive_state =
+    robot_ptr_->ReceiveMotionState(std::chrono::milliseconds(10));
 
-  if (!is_active_)
+  if ((msg_received_ = receive_state.return_code == kuka::external::control::ReturnCode::OK))
   {
-    std::this_thread::sleep_for(std::chrono::milliseconds(2));
-    msg_received_ = false;
-
-    return return_type::OK;
+    auto & req_message = robot_ptr_->GetLastMotionState();
+
+    std::copy(
+      req_message.GetMeasuredPositions()->begin(), req_message.GetMeasuredPositions()->end(),
+      hw_position_states_.begin());
+    std::copy(
+      req_message.GetMeasuredTorques()->begin(), req_message.GetMeasuredTorques()->end(),
+      hw_torque_states_.begin());
+    std::copy(
```

Comment:
is this necessary?

Reply:
very unlikely, I left this so it mirrors the previous behaviour. But we can leave this out